src/main/java/gift/dao/ProductDao.java

```diff
@@ -0,0 +1,106 @@
+package gift.dao;
+
+import gift.entity.Product;
+import gift.exception.DBServerException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.simple.JdbcClient;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public class ProductDao {
+    private final JdbcClient jdbcClient;
+
+    private static class ProductRowMapper implements RowMapper<Product> {
+        @Override
+        public Product mapRow(java.sql.ResultSet rs, int rowNum) throws java.sql.SQLException {
+            return new Product(
+                    rs.getLong("id"),
+                    rs.getString("name"),
+                    rs.getLong("price"),
+                    rs.getString("image_url")
+            );
+        }
+    }
+
+    public ProductDao(JdbcClient client) {;
+        this.jdbcClient = client;
+    }
+
+    public List<Product> findAll() {
+        String sql = "SELECT * FROM products";
+        return jdbcClient.sql(sql)
+                .query(new ProductRowMapper())
+                .stream()
+                .toList();
+    }
+
+    public Product findById(Long productId) {
+        String sql = "SELECT * FROM products WHERE id = ?";
+        return jdbcClient.sql(sql)
+                .param(productId)
+                .query(new ProductRowMapper())
+                .stream()
+                .findFirst()
+                .orElse(null);
+    }
+
+    public Long insertWithKey(Product product) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        String sql = "INSERT INTO products (name, price, image_url) VALUES (?, ?, ?)";
+        jdbcClient.sql(sql)
+            .param(product.getName())
+            .param(product.getPrice())
+            .param(product.getImageUrl())
+            .update(keyHolder);
+        if (keyHolder.getKey() == null) {
+            throw new DBServerException("상품을 저장하는 중 오류가 발생했습니다.(key에 제대로 생성되지 않았음)");
+        }
```

Comment:
예외처리를 하려고 했던 점은 너무 좋은 것 같습니다.
다만 keyHolder.getKey() == null 인 부분이, "DB 시스템 상의 문제"인건지, keyHolder에서 Key를 설정하지 못한 건지 등 정확한 이유를 알 수 없을 것 같다는 생각이 들어요.

예를 들어,
```suggestion
public Long insertWithKey(Product product) {
    KeyHolder keyHolder = new GeneratedKeyHolder();
    String sql = "INSERT INTO products (name, price, image_url) VALUES (?, ?, ?)";

    try {
        jdbcClient.sql(sql)
            .param(product.getName())
            .param(product.getPrice())
            .param(product.getImageUrl())
            .update(keyHolder);
    } catch (Exception e) {
        throw new DBServerException("DB Insert 중 예외가 발생했습니다.", e); // DB 서비스에서 발생한 예외라면 정확한 예외를 담아서 전파하는 게 좋기 때문에 exception을 추가해주는 게 좋아요.
    }

    Number key = keyHolder.getKey();
    if (key == null) {
        throw new DBServerException(String.format("상품 저장은 성공했지만, 생성된 키가 반환되지 않았습니다. 상품 정보: %s", product));
    }
...
}

```